File: proposed-projects/round-trip-experiment/reverse/20260909T023701Z_e94dca40/20260910T060603Z_391335fe/ported_implementation/gbnf/grammar_graph/grammar_graph_types.py
from __future__ import annotations

from enum import Enum
from typing import Union

from ..utils.validate_non_empty import validate_non_empty

# A Range is a two element list of code points, inclusive on both ends.
Range = list


class RuleType(str, Enum):
    CHAR = "char"
    CHAR_EXCLUDE = "char_exclude"
    END = "end"
    REF = "ref"

    def __str__(self) -> str:
        return str(self.value)


class RuleChar:
    def __init__(self, value: list | None = None) -> None:
        self.type = RuleType.CHAR
        self.value = list(value) if value is not None else []

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, RuleChar):
            return NotImplemented
        return self.value == other.value

    # Rules are compared by identity when used as keys, matching the reference
    # implementation's use of a Map keyed on the rule object.
    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return f"RuleChar({self.value!r})"


class RuleCharExclude:
    def __init__(self, value: list | None = None) -> None:
        self.type = RuleType.CHAR_EXCLUDE
        self.value = list(value) if value is not None else []

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, RuleCharExclude):
            return NotImplemented
        return self.value == other.value

    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return f"RuleCharExclude({self.value!r})"


class RuleEnd:
    def __init__(self) -> None:
        self.type = RuleType.END

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, RuleEnd):
            return NotImplemented
        return True

    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return "RuleEnd()"


UnresolvedRule = Union[RuleChar, RuleCharExclude, RuleEnd, "RuleRef"]

# ValidInput can either be a string, or a number indicating a code point.
# It CANNOT be a number representing a number; a number intended as input (like "8")
# should be passed in as a string.
ValidInput = Union[str, int, list]

# RuleRefs should never be exposed to the end user.
ResolvedRule = Union[RuleCharExclude, RuleChar, RuleEnd]

__all__ = [
    "Range",
    "ResolvedRule",
    "RuleChar",
    "RuleCharExclude",
    "RuleEnd",
    "RuleType",
    "UnresolvedRule",
    "ValidInput",
    "validate_non_empty",
]

File: packages/gbnf-experiment/data/20260909T181751Z_00b43b48/ported_implementation/src/grammar-graph/grammar-graph-types.ts
import { validateNonEmpty } from '../utils/validate-non-empty.js';
import type { GraphNode } from './graph-node.js';
import type { GraphPointer } from './graph-pointer.js';
import type { Pointers } from './pointers.js';
import { RuleRef } from './rule-ref.js';
import { RuleType } from './rule-type.js';

export { RuleRef, RuleType };

export type Colorize = (text: string | number, color: string) => string;

export interface PrintOpts {
  pointers?: Pointers;
  colorize: Colorize;
  showPosition?: boolean;
}

export type Range = [number, number];

export abstract class Rule {
  abstract readonly type: RuleType;
}

export abstract class RuleWithListOfIntsOrRanges extends Rule {
  value: (number | Range)[];

  constructor(value: (number | Range)[]) {
    super();
    // rules coming in are mutated as ranges and alternates are folded into them;
    // take a copy so that callers holding onto the original array are unaffected.
    this.value = validateNonEmpty(value).slice();
  }
}

export class RuleChar extends RuleWithListOfIntsOrRanges {
  readonly type = RuleType.CHAR;
}

export class RuleCharExclude extends RuleWithListOfIntsOrRanges {
  readonly type = RuleType.CHAR_EXCLUDE;
}

export class RuleEnd extends Rule {
  readonly type = RuleType.END;
}

export type UnresolvedRule = RuleChar | RuleCharExclude | RuleRef | RuleEnd;

// RuleRefs should never be exposed to the end user.
export type ResolvedRule = RuleCharExclude | RuleChar | RuleEnd;

// ValidInput can either be a string, or a number indicating a code point.
// It CANNOT be a number representing a number; a number intended as input (like "8")
// should be passed in as a string.
export type ValidInput = string | number | number[];

export type ResolvedGraphPointer = GraphPointer<ResolvedRule>;

export type GraphNodeRuleRef = GraphNode<RuleRef>;
